jupyterhub/spawner.py

```diff
@@ -119,50 +165,93 @@ def options_from_form(self, form_data):
+
+        Environment variables that end up in the single-user server's process come from 3 sources:
+          - This `environment` configurable
+          - The JupyterHub process' environment variables that are whitelisted in `env_keep`
+          - Variables to establish contact between the single-user notebook and the hub (such as JPY_API_TOKEN)
 
-        Value could be a string or a callable. If it is a callable, it will
-        be called with one parameter, which will be the instance of the spawner
-        in use. It should quickly (without doing much blocking operations) return
-        a string that will be used as the value for the environment variable.
+        The `enviornment` configurable should be set by JupyterHub administrators to add
+        installation specific environment variables. It is a dict where the key is the name of the environment
+        variable, and the value can be a string or a callable. If it is a callable, it will be called
+        with one parameter (the spawner instance), and should return a string fairly quickly (no blocking
+        operations please!).
+
+        Note that the spawner class' interface is not guaranteed to be exactly same across upgrades,
+        so if you are using the callable take care to verify it continues to work after upgrades!
         """
     ).tag(config=True)
-    
-    cmd = Command(['jupyterhub-singleuser'],
-        help="""The command used for starting notebooks."""
+
+    # FIXME: Add info about shell expansion here.
+    cmd = Command(
+        ['jupyterhub-singleuser'],
+        help="""
+        The command used for starting the single-user server.
+
+        You can provide this as either a list or a string. Note that this should only contain
+        the path to the script to start - extra arguments should be provided via `args`.
+
+        This is usually set if you want to start the single-user server in a different python
+        environment (with virtualenv/conda) than JupyterHub itself.
+        """
     ).tag(config=True)
-    args = List(Unicode(),
-        help="""Extra arguments to be passed to the single-user server"""
+
+    # FIXME: Add info about shell expansion here.
+    args = List(
+        Unicode(),
+        help="""
+        Extra arguments to be passed to the single-user server.
+
+        """
     ).tag(config=True)
-    
-    notebook_dir = Unicode('',
-        help="""The notebook directory for the single-user server
-        
-        `~` will be expanded to the user's home directory
-        `{username}` will be expanded to the user's username
+
+    notebook_dir = Unicode(
+        '',
+        help="""
+        Path to the notebook directory for the single-user server.
+
+        The user will see a file listing of this directory when they start, and the Notebook
+        interface itself will not allow browsing outside the contents of this directory (and its
```

Comment:
Path to the notebook directory for the single-user server.

The user sees a file listing of this directory when the notebook interface is started. The current interface does not easily allow browsing beyond the subdirectories in this directory's tree.